pkg/networking/middleware/response.go

```diff
@@ -0,0 +1,101 @@
+package middleware
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/snyk/error-catalog-golang-public/snyk"
+	"github.com/snyk/error-catalog-golang-public/snyk_errors"
+	"github.com/snyk/go-application-framework/pkg/configuration"
+	networktypes "github.com/snyk/go-application-framework/pkg/networking/network_types"
+)
+
+type ResponseMiddleware struct {
+	next       http.RoundTripper
+	config     configuration.Configuration
+	errHandler networktypes.ErrorHandlerFunc
+}
+
+func NewReponseMiddleware(roundTriper http.RoundTripper, config configuration.Configuration, errHandler networktypes.ErrorHandlerFunc) *ResponseMiddleware {
+	return &ResponseMiddleware{
+		next:       roundTriper,
+		config:     config,
+		errHandler: errHandler,
+	}
+}
+
+func (rm ResponseMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	res, err := rm.next.RoundTrip(req)
+
+	if err != nil {
+		return res, err
+	}
+
+	err = HandleResponse(res, rm.config)
+
+	if rm.errHandler != nil {
+		err = rm.errHandler(err, res.Request.Context())
+	}
+
+	// RoundTrip should return one or the other.
+	if err != nil {
+		res = nil
+	}
+
+	return res, err
+}
+
+// HandleResponse maps the response param to the eror catalog error.
+func HandleResponse(res *http.Response, config configuration.Configuration) error {
+	if res == nil {
+		return nil
+	}
+
+	apiUrl := config.GetString(configuration.API_URL)
+	additionalSubdomains := config.GetStringSlice(configuration.AUTHENTICATION_SUBDOMAINS)
+	additionalUrls := config.GetStringSlice(configuration.AUTHENTICATION_ADDITIONAL_URLS)
+	//nolint:errcheck // discarded error since it's returned only with the value of false
+	requiresAuth, _ := ShouldRequireAuthentication(apiUrl, res.Request.URL, additionalSubdomains, additionalUrls)
```

Comment:
Suggestion: Even with your comment, I think it is worth to use the error in the next condition. The reason is, that what you describe is the current behaviour, but there might be changes along the way that change this.